### PR TITLE
migration: refrain from offering all pages when possible

### LINKED ISSUE #387: Preserve modified page masks on migration sources
If a migration source calls `VmmHdl::track_dirty_pages` and migration subsequently fails, the source must try to transfer the tracked pages again if it tries to migrate again, even if the guest hasn't touched those pages again.

### PR DESCRIPTION
**Pull request by hawkw**
Presently, when transferring guest RAM as part of a migration, Propolis will
always offer *all* guest pages in the initial pre-pause RAM transfer phase.
After the guest is paused, Propolis will then offer any pages which were marked
as dirty since the pre-pause RAM transfer, ensuring the most recent guest memory
is transferred. In the initial pre-pause RAM push, however, we cannot offer only
dirty RAM, because tracking which pages are dirty using the
`VM_TRACK_DIRTY_PAGES` ioctl clears the dirty bit, and we may need to perform a
subsequent migration should this one fail. That subsequent migration will have a
bad time if it only offers dirty pages and we have cleared the dirty bits on any
pages that were touched only prior to the previous failed migration. This is a
shame, because we may end up transferring a bunch of pages that the guest has
never actually touched and doesn't care about.

Now, thanks to @pfmooney's upstream work in [illumos-gate#3266], we can
be more efficient about what pages are transferred in the initial push, without
dropping a bunch of pages on the ground if we retry a failed migration. Using
the `VNO_OP_SET_DIRTY` operation performed by the `VM_NPT_OPERATION` ioctl that
Patrick has added to bhyve, we can put back any dirty bits that were cleared
during a failed migration, allowing a subsequent migration to also offer only
dirty pages. Since this ioctl is only available with the latest bhyve (v17), we
will still fall back to the previous behavior of offering all guest RAM if
 `VM_NPT_OPERATION` is unavailable.

Initially, I had planned to just change this code to use the `VNO_OP_GET_DIRTY`
operation, which accesses but does not clear the dirty bits. This way, I
figured, we could just always transfer all dirty pages. However, we *do*
actually want to be clearing dirty bits initially, because we want the
post-pause RAM push to only have to transfer pages that have been touched
_since_ the pre-pause RAM push phase. If we used `VNO_OP_GET_DIRTY` to read the
dirty bits without clearing them for the initial pre-pause RAM push, the
post-pause RAM push would offer all the same pages, which defeats the purpose of
splitting the RAM push into two pages in the first place. Thus, instead, we do
clear the dirty bits, but we store which pages were dirty and put them back
using `VNO_OP_SET_DIRTY` should the migration fail.

The PHD tests I added way back in #623 test retrying a failed migration with
running processes, which should exercise this change. I've run the tests on
`atrium`, which has an Illumos kernel that includes Patrick's bhyve changes:

<details open>

<summary>bhyve version on `atrium`</summary>

```console
eliza@atrium ~ $ pkg info system/bhyve
             Name: system/bhyve
          Summary: BSD hypervisor
      Description: BSD hypervisor
         Category: System/Virtualization
            State: Installed (Manually installed)
        Publisher: helios-dev
          Version: 0.5.11
           Branch: 2.0.22550
   Packaging Date: Wed Mar  6 17:51:32 2024
Last Install Time: Thu Dec  2 05:52:30 2021
 Last Update Time: Thu Mar  7 17:31:21 2024
             Size: 1.48 MB
             FMRI: pkg://helios-dev/system/bhyve@0.5.11-2.0.22550:20240306T175132Z
```

</details>
<details>
<summary>We can see the first RAM push for the first migration, which will fail:</summary>

```log
23:37:54.890Z INFO propolis-server (vm_controller): ram_push (RamPushPrePause): got query for range 0x0..0xffffffffffffffff, vm range GuestAddr(
        0x0,
    )..=GuestAddr(
        0x1fffffff,
    )
23:37:54.890Z INFO propolis-server (vm_controller): offering ram
    can_redirty_pages = true
    discipline = OfferDirty
23:37:54.895Z INFO propolis-server (vm_controller): ram_push: offering 16041 pages between 0x0 and 0x8000000
23:37:54.898Z INFO propolis-server (vm_controller): ram_push: offering 0 pages between 0x8000000 and 0x10000000
23:37:54.901Z INFO propolis-server (vm_controller): ram_push: offering 0 pages between 0x10000000 and 0x18000000
23:37:54.904Z INFO propolis-server (vm_controller): ram_push: offering 0 pages between 0x18000000 and 0x20000000
23:37:54.904Z INFO propolis-server (vm_controller): ram_push: xfer RAM between 0x0 and 0x8000000
23:37:56.312Z INFO propolis-server (vm_controller): ram_push: done sending ram
23:37:56.313Z INFO propolis-server (vm_controller): Pausing devices
23:37:56.313Z INFO propolis-server (vcpu_tasks): vCPU paused
    vcpu = 0
23:37:56.313Z INFO propolis-server (vcpu_tasks): vCPU paused
    vcpu = 1
23:37:56.313Z INFO propolis-server (vm_controller): Sending pause request to com1
23:37:56.313Z INFO propolis-server (vm_controller): Sending pause request to lpc-bhyve-hpet
23:37:56.313Z INFO propolis-server (vm_controller): Sending pause request to lpc-ps2ctrl
23:37:56.313Z INFO propolis-server (vm_controller): Sending pause request to pci-i440fx-hb
23:37:56.313Z INFO propolis-server (vm_controller): Sending pause request to pci-nvme-0.4.0
23:37:56.313Z INFO propolis-server (vm_controller): Sending pause request to pci-piix3-lpc
23:37:56.313Z INFO propolis-server (vm_controller): Sending pause request to pci-piix3-pm
23:37:56.313Z INFO propolis-server (vm_controller): Sending pause request to qemu-fwcfg
23:37:56.313Z INFO propolis-server (vm_controller): Sending pause request to qemu-lpc-debug
23:37:56.313Z INFO propolis-server (vm_controller): Sending pause request to qemu-ramfb
23:37:56.313Z INFO propolis-server (vm_controller): Sending pause request to test-migration-failure
23:37:56.313Z INFO propolis-server (vm_controller): Sending pause request to vcpu-0
23:37:56.313Z INFO propolis-server (vm_controller): Sending pause request to vcpu-1
23:37:56.313Z INFO propolis-server (vm_controller): Waiting for devices to pause
23:37:56.313Z INFO propolis-server (vm_controller): Got paused future from dev com1
23:37:56.313Z INFO propolis-server (vm_controller): Got paused future from dev lpc-bhyve-hpet
23:37:56.313Z INFO propolis-server (vm_controller): Got paused future from dev lpc-ps2ctrl
23:37:56.313Z INFO propolis-server (vm_controller): Got paused future from dev pci-i440fx-hb
23:37:56.313Z INFO propolis-server (vm_controller): Got paused future from dev pci-nvme-0.4.0
23:37:56.313Z INFO propolis-server (vm_controller): Got paused future from dev pci-piix3-lpc
23:37:56.313Z INFO propolis-server (vm_controller): Got paused future from dev pci-piix3-pm
23:37:56.313Z INFO propolis-server (vm_controller): Got paused future from dev qemu-fwcfg
23:37:56.313Z INFO propolis-server (vm_controller): Got paused future from dev qemu-lpc-debug
23:37:56.313Z INFO propolis-server (vm_controller): Got paused future from dev qemu-ramfb
23:37:56.313Z INFO propolis-server (vm_controller): Got paused future from dev test-migration-failure
23:37:56.313Z INFO propolis-server (vm_controller): Got paused future from dev vcpu-0
23:37:56.313Z INFO propolis-server (vm_controller): Got paused future from dev vcpu-1
23:37:56.313Z INFO propolis-server (vm_controller): dev com1 completed pause
23:37:56.313Z INFO propolis-server (vm_controller): dev lpc-bhyve-hpet completed pause
23:37:56.313Z INFO propolis-server (vm_controller): dev lpc-ps2ctrl completed pause
23:37:56.313Z INFO propolis-server (vm_controller): dev pci-i440fx-hb completed pause
23:37:56.313Z INFO propolis-server (vm_controller): dev pci-nvme-0.4.0 completed pause
23:37:56.313Z INFO propolis-server (vm_controller): dev pci-piix3-lpc completed pause
23:37:56.313Z INFO propolis-server (vm_controller): dev pci-piix3-pm completed pause
23:37:56.313Z INFO propolis-server (vm_controller): dev qemu-fwcfg completed pause
23:37:56.313Z INFO propolis-server (vm_controller): dev qemu-lpc-debug completed pause
23:37:56.313Z INFO propolis-server (vm_controller): dev qemu-ramfb completed pause
23:37:56.313Z INFO propolis-server (vm_controller): dev test-migration-failure completed pause
23:37:56.313Z INFO propolis-server (vm_controller): dev vcpu-0 completed pause
23:37:56.313Z INFO propolis-server (vm_controller): dev vcpu-1 completed pause
23:37:56.313Z INFO propolis-server (vm_controller): all devices paused
23:37:56.313Z INFO propolis-server (vm_controller): Pausing kernel VMM resources
23:37:56.313Z INFO propolis-server (vm_controller): ram_push (RamPushPostPause): got query for range 0x0..0xffffffffffffffff, vm range GuestAddr(
        0x0,
    )..=GuestAddr(
        0x1fffffff,
    )
23:37:56.314Z INFO propolis-server (vm_controller): offering ram
    can_redirty_pages = true
    discipline = OfferDirty
23:37:56.317Z INFO propolis-server (vm_controller): ram_push: offering 95 pages between 0x0 and 0x8000000
23:37:56.321Z INFO propolis-server (vm_controller): ram_push: offering 0 pages between 0x8000000 and 0x10000000
23:37:56.326Z INFO propolis-server (vm_controller): ram_push: offering 0 pages between 0x10000000 and 0x18000000
23:37:56.330Z INFO propolis-server (vm_controller): ram_push: offering 0 pages between 0x18000000 and 0x20000000
23:37:56.330Z INFO propolis-server (vm_controller): ram_push: xfer RAM between 0x0 and 0x8000000
23:37:56.339Z INFO propolis-server (vm_controller): ram_push: done sending ram
```
</details>
<summary>The test migration failure device fails to be imported, so the
source Propolis gives up on the migration, and we put back the dirty bits:
</summary>

```log
23:37:56.381Z ERRO propolis-server (vm_controller): remote error: failed to migrate device state: failed to apply deserialized device state: you have no chance to survive, make your time
23:37:56.383Z INFO propolis-server (vm_controller): re-dirtied 16136 pages at 0x0
23:37:56.383Z ERRO propolis-server (vm_controller): Migration task failed: Destination migration instance encountered error: failed to migrate device state: failed to apply deserialized device state: you have no chance to survive, make your time
23:37:56.383Z INFO propolis-server (vm_state_worker): Migration source task exited: Err(RemoteError(Destination, "failed to migrate device state: failed to apply deserialized device state: you have no chance to survive, make your time"))
23:37:56.383Z INFO propolis-server (vm_controller): Resuming kernel VMM resources
```
</details>
<details>
<summary>
A second migration attempt transfers all the expected pages:
</summary>

```log
23:37:57.759Z INFO propolis-server (vm_controller): offering ram
    can_redirty_pages = true
    discipline = OfferDirty
23:37:57.767Z INFO propolis-server (vm_controller): ram_push: offering 16041 pages between 0x0 and 0x8000000
23:37:57.774Z INFO propolis-server (vm_controller): ram_push: offering 0 pages between 0x8000000 and 0x10000000
23:37:57.777Z INFO propolis-server (vm_controller): ram_push: offering 0 pages between 0x10000000 and 0x18000000
23:37:57.780Z INFO propolis-server (vm_controller): ram_push: offering 0 pages between 0x18000000 and 0x20000000
23:37:57.780Z INFO propolis-server (vm_controller): ram_push: xfer RAM between 0x0 and 0x8000000
23:37:59.187Z INFO propolis-server (vm_controller): ram_push: done sending ram
23:37:59.187Z INFO propolis-server (vm_controller): Pausing devices
23:37:59.187Z INFO propolis-server (vcpu_tasks): vCPU paused
    vcpu = 0
23:37:59.188Z INFO propolis-server (vcpu_tasks): vCPU paused
    vcpu = 1
23:37:59.188Z INFO propolis-server (vm_controller): Sending pause request to com1
23:37:59.188Z INFO propolis-server (vm_controller): Sending pause request to lpc-bhyve-hpet
23:37:59.188Z INFO propolis-server (vm_controller): Sending pause request to lpc-ps2ctrl
23:37:59.188Z INFO propolis-server (vm_controller): Sending pause request to pci-i440fx-hb
23:37:59.188Z INFO propolis-server (vm_controller): Sending pause request to pci-nvme-0.4.0
23:37:59.188Z INFO propolis-server (vm_controller): Sending pause request to pci-piix3-lpc
23:37:59.188Z INFO propolis-server (vm_controller): Sending pause request to pci-piix3-pm
23:37:59.188Z INFO propolis-server (vm_controller): Sending pause request to qemu-fwcfg
23:37:59.188Z INFO propolis-server (vm_controller): Sending pause request to qemu-lpc-debug
23:37:59.188Z INFO propolis-server (vm_controller): Sending pause request to qemu-ramfb
23:37:59.188Z INFO propolis-server (vm_controller): Sending pause request to test-migration-failure
23:37:59.188Z INFO propolis-server (vm_controller): Sending pause request to vcpu-0
23:37:59.188Z INFO propolis-server (vm_controller): Sending pause request to vcpu-1
23:37:59.188Z INFO propolis-server (vm_controller): Waiting for devices to pause
23:37:59.188Z INFO propolis-server (vm_controller): Got paused future from dev com1
23:37:59.188Z INFO propolis-server (vm_controller): Got paused future from dev lpc-bhyve-hpet
23:37:59.188Z INFO propolis-server (vm_controller): Got paused future from dev lpc-ps2ctrl
23:37:59.188Z INFO propolis-server (vm_controller): Got paused future from dev pci-i440fx-hb
23:37:59.188Z INFO propolis-server (vm_controller): Got paused future from dev pci-nvme-0.4.0
23:37:59.188Z INFO propolis-server (vm_controller): Got paused future from dev pci-piix3-lpc
23:37:59.188Z INFO propolis-server (vm_controller): Got paused future from dev pci-piix3-pm
23:37:59.188Z INFO propolis-server (vm_controller): Got paused future from dev qemu-fwcfg
23:37:59.188Z INFO propolis-server (vm_controller): Got paused future from dev qemu-lpc-debug
23:37:59.188Z INFO propolis-server (vm_controller): Got paused future from dev qemu-ramfb
23:37:59.188Z INFO propolis-server (vm_controller): Got paused future from dev test-migration-failure
23:37:59.188Z INFO propolis-server (vm_controller): Got paused future from dev vcpu-0
23:37:59.188Z INFO propolis-server (vm_controller): Got paused future from dev vcpu-1
23:37:59.188Z INFO propolis-server (vm_controller): dev com1 completed pause
23:37:59.188Z INFO propolis-server (vm_controller): dev lpc-bhyve-hpet completed pause
23:37:59.188Z INFO propolis-server (vm_controller): dev lpc-ps2ctrl completed pause
23:37:59.188Z INFO propolis-server (vm_controller): dev pci-i440fx-hb completed pause
23:37:59.188Z INFO propolis-server (vm_controller): dev pci-nvme-0.4.0 completed pause
23:37:59.188Z INFO propolis-server (vm_controller): dev pci-piix3-lpc completed pause
23:37:59.188Z INFO propolis-server (vm_controller): dev pci-piix3-pm completed pause
23:37:59.188Z INFO propolis-server (vm_controller): dev qemu-fwcfg completed pause
23:37:59.188Z INFO propolis-server (vm_controller): dev qemu-lpc-debug completed pause
23:37:59.188Z INFO propolis-server (vm_controller): dev qemu-ramfb completed pause
23:37:59.188Z INFO propolis-server (vm_controller): dev test-migration-failure completed pause
23:37:59.188Z INFO propolis-server (vm_controller): dev vcpu-0 completed pause
23:37:59.188Z INFO propolis-server (vm_controller): dev vcpu-1 completed pause
23:37:59.188Z INFO propolis-server (vm_controller): all devices paused
23:37:59.189Z INFO propolis-server (vm_controller): Pausing kernel VMM resources
23:37:59.189Z INFO propolis-server (vm_controller): ram_push (RamPushPostPause): got query for range 0x0..0xffffffffffffffff, vm range GuestAddr(
        0x0,
    )..=GuestAddr(
        0x1fffffff,
    )
23:37:59.189Z INFO propolis-server (vm_controller): offering ram
    can_redirty_pages = true
    discipline = OfferDirty
23:37:59.191Z INFO propolis-server (vm_controller): ram_push: offering 89 pages between 0x0 and 0x8000000
23:37:59.194Z INFO propolis-server (vm_controller): ram_push: offering 0 pages between 0x8000000 and 0x10000000
23:37:59.197Z INFO propolis-server (vm_controller): ram_push: offering 0 pages between 0x10000000 and 0x18000000
23:37:59.200Z INFO propolis-server (vm_controller): ram_push: offering 0 pages between 0x18000000 and 0x20000000
23:37:59.200Z INFO propolis-server (vm_controller): ram_push: xfer RAM between 0x0 and 0x8000000
23:37:59.208Z INFO propolis-server (vm_controller): ram_push: done sending ram
```
</details>

Closes #387

[illumos-gate#3266] https://code.illumos.org/c/illumos-gate/+/3266

### CODE REVIEW

File: bin/propolis-server/src/lib/migrate/source.rs
Comment (dancrossnyc) [inline bin/propolis-server/src/lib/migrate/source.rs:467]: Is there an edge case here in that, if the region size does not start or end on a page address that is a multiple of 8 pages, this will attempt to send pages that aren't in the region?

File: bin/propolis-server/src/lib/migrate/source.rs
Comment (dancrossnyc) [inline bin/propolis-server/src/lib/migrate/source.rs:217]: This representation seems overly complex, and moreover, perhaps not as efficient as one might think.

Consider that each tuple in the vec will almost certainly be aligned on a 16-byte boundary, meaning that 256 entries will take more than 4KiB of host memory, including overhead from the Vec. Combined with the overhead of the hash map, I can see some scenarios when this may be slightly more compact, but I can also see many where it's not.  It's really only a win in cases where modifications to the address space are extremely sparse, and it's not clear to me that that's the common case.

I think that, in absence of evidence that memory is unduly affected by storing bitmaps for resetting dirty bits, I'd just use a bitmap.  Recall also that that's an incredibly efficient representation; since each 4KiB page covers 128 MiB of address space.

I'd certainly acquiesce in my objection if there were some measurements that showed it to be a win on average?

File: bin/propolis-server/src/lib/migrate/source.rs
Comment (dancrossnyc) [inline bin/propolis-server/src/lib/migrate/source.rs:204]: It would be useful to mention that dirty bits are accumulated across all phases of memory transfer.

File: bin/propolis-server/src/lib/migrate/source.rs
Comment (dancrossnyc) [inline bin/propolis-server/src/lib/migrate/source.rs:71]: There is a third: scan each page and refrain from sending it if it's all zeroes. This may be a useful optimization in any event, as many "dirty" pages were probably just zeroed by the guest. Of course, it trades off CPU time for reduced latency and using less memory for maintaining the accumulated set of dirty pages, so that's got to be balanced against the requirement of not starving running guests, but considering that dirty pages can short-circuit the scan as soon as they find a non-zero byte (or u64) it may not be that bad.

File: bin/propolis-server/src/lib/migrate/source.rs
Comment (dancrossnyc) [inline bin/propolis-server/src/lib/migrate/source.rs:245]: Why make this a static, and not, say, an attribute on the machine or something like that?

Comment (hawkw) [reply bin/propolis-server/src/lib/migrate/source.rs:245]: My thinking was that, because it's state related to the migration source implementation in particular, it ought to be scoped to this module, rather than leaked out into the machine type. But, you're right that the static is a bit weird — even though a `propolis-server` process will only ever manage one VM, this module isn't really written with that assumption, and instead stores all the state on the machine type. I'm happy to move this if you prefer!

File: bin/propolis-server/src/lib/migrate/source.rs
Comment (pfmooney) [reply bin/propolis-server/src/lib/migrate/source.rs:217]: > this code is assuming that guest address spaces are sparser than they are in real life

Yeah.  Save for guests which are migrated soon after they are provisioned, I would be shocked if any instance which had been up for a day or more (probably hours even) left any page un-dirtied.

So while it will be nice to cut down on migration time/cost for freshly provisioned instances, we probably shouldn't over-rotate on optimizing for that case.

File: bin/propolis-server/src/lib/migrate/source.rs
Comment (hawkw) [reply bin/propolis-server/src/lib/migrate/source.rs:245]: Yeah, that makes sense. I'll add a struct in this module for migration-specific bits and add it to the machine state. Thanks!

File: bin/propolis-server/src/lib/migrate/source.rs
Comment (hawkw) [reply bin/propolis-server/src/lib/migrate/source.rs:467]: Well, so, there is a potential edge case, but it won't actually result in us
attempting to send more pages. `pages_offered` doesn't actually _control_ what pages we
offer, instead, we just check if it's non-zero to determine whether to offer
anything. The integer value is only used for logging. So, we may log that we are
offering more pages than we actually are...

Comment (hawkw) [reply bin/propolis-server/src/lib/migrate/source.rs:467]: I believe that `make_mem_offer` already handles the case where the bitmap is
larger than the actual number of pages in the region. So, perhaps the logging
should just happen there, instead.

File: bin/propolis-server/src/lib/migrate/source.rs
Comment (dancrossnyc) [inline bin/propolis-server/src/lib/migrate/source.rs:232]: Nit: I know it's totally legal, but it surprises me a bit to see the constant defined on the line _after_ it's used. :-D

Comment (hawkw) [reply bin/propolis-server/src/lib/migrate/source.rs:232]: happy to move it if you'd like

Comment (dancrossnyc) [reply bin/propolis-server/src/lib/migrate/source.rs:232]: I leave it to your capable discretion. It's just always a bit surprising to me when I encounter it.

File: bin/propolis-server/src/lib/migrate/source.rs
Comment (dancrossnyc) [reply bin/propolis-server/src/lib/migrate/source.rs:455]: Okee doke.

File: bin/propolis-server/src/lib/migrate/source.rs
Comment (dancrossnyc) [reply bin/propolis-server/src/lib/migrate/source.rs]: This begs the question: is this optimization worth it, then? If the average case is that O(all) pages have been touched in a fairly short time after the VM starts, and we don't expect migrations to happen _that_ frequently, perhaps it's best to just stick with the simpler approach?

**Comment by dancrossnyc** [comment]
I think this change looks fine to me, but speaking for myself, I'd only feel comfortable with it being submitted after getting Patrick's approval, as he has way more state than I do.

**Review by dancrossnyc — APPROVED** [review]

**Comment by hawkw** [comment]
> I'm curious about what your testing plan is for this, prior to integration.

As I discussed in the PR description, the PHD tests added in https://github.com/oxidecomputer/propolis/pull/623 were intended to exercise this. Those tests migrate a running process that has dirtied a few pages in the guest's address space, and then resume that process post-migration and let it try to read from its memory again. If there's anything else you think would be necessary to test this change, I'd love your advice!

File: bin/propolis-server/src/lib/migrate/source.rs
Comment (hawkw) [reply bin/propolis-server/src/lib/migrate/source.rs:171]: This could probably be `debug`, I'll change it.

**Comment by pfmooney** [comment]
> As I discussed in the PR description, the PHD tests added in #623 were intended to exercise this.

Please pardon me for missing that.

I assume that in your testing, you observed the logging evidence of the re-dirtying code path being taken in lieu of the prior send-it-all approach?  It would be good to toss any notes like that in a comment on the PR for posterity.

**Comment by hawkw** [comment]
> I assume that in your testing, you observed the logging evidence of the re-dirtying code path being taken in lieu of the prior send-it-all approach? It would be good to toss any notes like that in a comment on the PR for posterity.

I also included some logs in the PR description. :)

**Review by pfmooney — APPROVED** [review]
Thanks for the additional testing details.

**Pull request merged**
